### ISSUE TITLE
Way to disable resizing, or improve resizing quality

### ISSUE THREAD

**Issue by baoshan** [desbma/sacad#6]
Hi, I'm not sure if there're some filter/enhancement when `sacad` resize the images.

Please compare:

`sacad "Alice Sara Ott" "Pictures" 1400 output.jpg`

with

`sacad -t 0 "Alice Sara Ott" "Pictures" 1400 output.jpg`

The resized image has a zigzag text border.

For archiving purpose, I think an option to preserve the original image is necessary. Or can `sacad` be smarter by finding **the** best quality picture?

Thanks.

**Comment by desbma** [comment]
Interesting example, thanks for reporting it.

They are 3 factors that affect quality in this case:

1. When resizing, the algorithm used can change quality dramatically
2. After resizing SACAD applies a filter to sharpen a slightly the image, to compensate for the typical blur introduced by downsizing.
3. The image is recompressed (because the 16 images have to be joined into 1). JPEG is lossy, so that means some data is lost, although it is probably not perceptible because SACAD recompresses with a high quality (higher than most source image compression).

In this example changing 1 can improve quality significantly, so I'll definitely look into this.

**Comment by desbma** [comment]
Original:
[![original image](https://i.imgur.com/pA9MEl7s.jpg)](https://i.imgur.com/pA9MEl7.jpg)
Resized with current method:
[![resized old](https://i.imgur.com/2y6zQOis.jpg)](https://i.imgur.com/2y6zQOi.jpg)
Resized, with a new method:
[![resized improved](https://i.imgur.com/Tehf5LFs.jpg)](https://i.imgur.com/Tehf5LF.jpg)

**Comment by baoshan** [comment]
That's indeed an improvement. Looks even better (yes, perceived) than the original (low quality encoder, I guess) one. :)

I want to pitch the idea of matching the highest quality again:

`SACAD is a smart cover art downloader, it can find the highest quality cover art for your collection.`

Once the highest quality cover is found, users can use `--resize` option to (optional) resize the image to specified size.

`sacad "Alice Sara Ott" "Pictures" pictures.jpg` and `sacad --resize 1200 "Alice Sara Ott" "Pictures" pictures.jpg`

This will lead to:

1. Simpler command line interface (can't be more simpler if we stick to the album / artist scheme);
2. For users specify the `--resize` option, he/she will get a unified image size (I know the covers are not always square);
3. Best quality image (I know it depends on the winner algorithm, but it seems sacad can make it).

What do you think about it?

**Comment by desbma** [comment]
You missed an important thing: the size parameter is not just used for resizing at the end, it is also used as a search and sorting factor to choose the result that will be downloaded.

I serves a purpose, take this use case for example: I want to download album cover for my music on my phone which has a screen size of 960x540. Covers are square so I target 540x540 covers. I don't need more, because that would take useless space, and waste CPU power to display images to the screen resolution (also depending on the app, the resizing may be crappy).
Let's say I get these results:
* 600x600 PNG cover on Last.fm
* 1400x1400 JPEG on Amazon

What do you think is the best result for the requested size? SACAD will choose the first one because:
* It's PNG which is better than JPEG quality wise
* Very close to target size, so no need to downsize (which would mean some loss for JPEG)
* Last.fm is more reliable than Amazon (which for example can have badges over album covers)

So depending on the parameters, the highest resolution is not always the deciding factor. If it is what you want, you can run for example `sacad "Alice Sara Ott" "Pictures" 1400 -t 50` which won't resize anything below 2100px (1400 + 1400 * 0.5) wide. 
There is an exception if the result comes from the Amazon Digital source as several subimages. The result will have the target 1400x1400 size because a new image will be generated anyway, so the JPEG recompression step can not be avoided and there is no downside to resizing in the same step.

**Comment by baoshan** [comment]
Thanks for sharing your thoughts!

>  the size parameter is not just used for resizing at the end, it is also used as a search and sorting factor to choose the result that will be downloaded.

I didn't miss that. Could we talk about the design first, not the implementation. So if you believe the direction is promising, we can find way to make it.

> take this use case for example

That's a good scenario!

> Covers are square so I target 540x540 covers. I don't need more, because that would take useless storage space, and waste CPU power to display images to the screen resolution, (also depending on the app, the resizing may be crappy)

Make sense, totally.

> What do you think is the best result for the requested size? SACAD will choose the first one because:

That's not my choice.

## PNG vs JPG

PNG itself is lossless, but where does the actual content come from? Given only `last.fm` provides `600 by 600` PNG, do you believe it's downscaled from the original TIF file, or a JPG file? I'm a pessimist.

## Size

Either choice leaves the actual downscaling (600 to 540 or 1440 to 540) to the app (or browser), which has a downscaling algorithm unknown, which can be crappy. So if you're set to a `960x540` screen, you'd better control the image to be `540` exactly at your side. (I assume you want to use many jukebox softwares),

Resizing a image from `600 by 600` to `540 by 540` will be crappier than resizing from `1600 by 1600`, for most algorithm (including `sacad`). Mathematically, image downscaling is the same as audio downscaling. Why many rebook cd indicating they're downscaled from `96Khz` or `192Khz`?

Secondly, the `600 x 600` image is a downscaled image from a master image (maybe the `1600 by 1600` one) probably. Downscale multiple times will introduced perceived quality degrade. You can find real examples [here](http://support.softartisans.com/kbview_1308.aspx) and [here](http://graphicdesign.stackexchange.com/questions/1672/will-scaling-down-incrementally-hurt-quality).

> So depending on the parameters, the highest resolution is not always the best result. If it is what you want, you can run for example sacad "Alice Sara Ott" "Pictures" 1400 -t 50 which won't resize anything below 2100px (1400 + 1400 * 0.5) wide. 

Increase the tolerance can also find me a 800px image, instead of the 2100px. Correct?

**Comment by desbma** [comment]
>  PNG itself is lossless, but where does the actual content come from? Given only `last.fm` provides `600 by 600` PNG, do you believe it's downscaled from the original TIF file, or a JPG file? I'm a pessimist.

Maybe you are right, maybe not. Why would Last.fm expose PNG images and waste bandwidth (which means money) if they come from a lossy source? Unless you have proof that is the case, it makes sense to favor PNG images in the algorithm.

> Resizing a image from `600 by 600` to `540 by 540` will be crappier than resizing from `1600 by 1600`, for most algorithm (including `sacad`). Mathematically, image downscaling is the same as audio downscaling. Why many red-book cd indicating they're downscaled from `96Khz` or `192Khz`?

I'm not sure this is true. The audio analogy is interesting but beware, so called *audiophile* communities spread a lot of affirmations that have absolutely no scientific ground, like all the fuss about [high res audio that is just a big hoax](https://people.xiph.org/~xiphmont/demo/neil-young.html).
For audio resampling, I you use a high quality resampler like [SoX](http://sox.sourceforge.net/SoX/Resampling), the downsampling will not introduce audible artifacts no watter if you do 48KHz -> 44.1KHz, 192KHz -> 48KHz or whatever. The claim you refer to is probably just a marketing thing.

> Increase the tolerance can also find me a 800px image, instead of the 2100px. Correct?

Right. But only if no better image was found. So it's a smaller image instead of no image.
It this annoys you, I could add a `--never-resize` parameter or similar, but I don't see a practical use case for this.

**Comment by baoshan** [comment]
> Unless you have proof that is the case, it makes sense to favor PNG images in the algorithm.

I can't find even one cover art in png format from [last.fm](http://last.fm). Even those uploaders (last.fm users) marked as `png`, last.fm serves them in `jpg` format. I'm not familiar with last.fm, please correct me if that's not true. If we have some new cover art (2010+) in png format, we can make do some quality study.

If both last.fm and amazon serves jpg. And we can not scoring the quality objectively (I think we can, actually). Based on their business model, amazon use merchants provided images mostly.

Can you find the below 2015 best-selling from last.fm api and make a comparison? Resolution and file size can make a rationale judgement?

http://z-ecx.images-amazon.com/images/P/B00MRHANNI.09._SCRMZZZZZZ_.jpg
http://z-ecx.images-amazon.com/images/P/B013CWFOR6.09._SCRMZZZZZZ_.jpg
http://z-ecx.images-amazon.com/images/P/B016WW10SA.09._SCRMZZZZZZ_.jpg

> The audio analogy is interesting but beware, so called audiophile communities spread a lot of affirmations that have absolutely no scientific ground.

Sampling at 192Khz makes the (digital) low-pass-filter design much easy. There're buzzwords around high res. But that has nothing to do with the mathematical root. I'm not a professional here. But I can definitely confirm the that's mathematically based.

For image downsizing, I can see [here](http://support.softartisans.com/kbview_1308.aspx) and [here](http://graphicdesign.stackexchange.com/questions/1672/will-scaling-down-incrementally-hurt-quality) clearly shows multiple-times downscaling hurts the quality.

Back to our topic, downsizing has far less impact on quality than the origin of the source. My main point here is:

__Larger res image is more likely provided by labels, small resolution is more likely provided by users upload.__ Which is the first factor impacting image quality. There're large res images actually is scanned (not authentically digital), but users do that mostly due to lack of high-quality cover art.

Thanks, again.

**Comment by desbma** [comment]
> I can't find even one cover art in png format from [last.fm](http://last.fm).

They are almost always PNG, and very very rarely GIF (I've seen this only once). And they obviously don't come from a scan.
Check with `https://ws.audioscrobbler.com/2.0/?method=album.getinfo&api_key=2410a53db5c7490d0f50c100a020f359&album=REPLACE_WITH_ALBUM&artist=REPLACE_WITH_ARTIST`

> Sampling at 192Khz makes the (digital) low-pass-filter design much easy.

'easy' has nothing to do with quality.

> For image downsizing, I can see here and here clearly shows multiple-times downscaling hurts the quality.

We agree that several downsizing steps clearly degrades quality.
But when downsizing (or downsampling for audio) with a sufficiently high quality algorithm, the source size (or sample rate) does not matter.

> __Larger res image is more likely provided by labels, small resolution is more likely provided by users upload.__ Which is the first factor impacting image quality. There're large res images actually is scanned (not authentic digital), but users do that mostly due to lack of high-quality cover art.

This is speculation not backed up with facts. I've never seen a single scanned cover on Last.fm.
The CoverLib source (now defunct) also had covers that were scanned at very high resolutions (so high you could see the grain and imperfections of the paper).

We could discuss theories for a long time, but back to the original issue: I have improved a lot resizing quality (which will be available in the next version).

If you believe SACAD makes the wrong choice of cover for a query, please open another issue and be specific about what is wrong.

If you want to suggest a feature, please open a new issue and describe the use case that SACAD does not satisfy, because I don't clearly see what you want yet.

**Comment by baoshan** [comment]
> They are almost always PNG, and very very rarely GIF (I've seen this only once). And they obviously don't come from a scan.

Thanks for the information. Last.fm doesn't serve a single `png` for their own website. But their API uses `png` format. It's easier now for comparison. I'll use the above 3 albums for a fair comparison.

> 'easy' has nothing to do with quality.

It looks like you're not interested with the mathematically proven lower THD, which can also be instrument measured. Theory and instrument always go beyond human perception. But for cover art selection, I'll try to persuade you by real cases.

> If you believe SACAD makes the wrong choice of cover for a query

Yes, I believe. Will report later in another thread. Thanks.
